pkg/logging/logger.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"context"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/provider"
+)
+
+// Logger wraps a time-ordered log set to allow batched logging of unique messages.
+type Logger struct {
+	messages *TimeOrderedLogSet
+	index    int
+	ctx      context.Context
+	host     *provider.HostClient
+	urn      resource.URN
+}
+
+// NewLogger returns an initialized Logger.
+func NewLogger(ctx context.Context, host *provider.HostClient, urn resource.URN) *Logger {
+	return &Logger{
+		messages: &TimeOrderedLogSet{},
+		ctx:      ctx,
+		host:     host,
+		urn:      urn,
+	}
+}
+
+// LogMessage adds a message to the log set and flushes the queue to the host.
+func (l *Logger) LogMessage(severity diag.Severity, s string) {
+	l.EnqueueMessage(severity, s)
+	l.LogNewMessages()
+}
+
+// EnqueueMessage adds a message to the log set but does not log it to the host.
+func (l *Logger) EnqueueMessage(severity diag.Severity, s string) {
+	l.messages.Add(Message{s, severity})
+}
+
+// GetNewMessages returns the list of new messages since last calling GetNewMessages.
+// Note: this method is not thread-safe, but is currently only called from a single process
+// because a new Logger is spawned for each resource.
+func (l *Logger) GetNewMessages() []Message {
+	idx := l.index
```

Comment:
Yeah, I'll defer that to another PR, since current usage is already one-logger-per-thread.

Most of the current log messages are status messages generated by the provider, and don't have a unique ID associated with them.

Reply:
I actually think that we should make this thread-safe since it seems so easy and protects against the future.

Reply:
Ok, will do.